library("ggplot2")
library("readxl")
library("dplyr")
library("tidyverse")
library("ggpubr")
library("ape")
library("lme4")
library("gplots")
library("plotly")
library("tidyr")
library("vegan")
library("data.table")
library("stringr")

#the goal here is to calculate p-values for the significant differences between flux pathways to use for correlating with gene expression
#differences. Also, we want to abundance filter flux pathways that do not appear in at least half the number of samples 
#(before running stats on identifying fluxes with significantly different activity between tumor and normal samples).

#loosely based on 05.27.21_microbial_inputoutput_analysis_Burns_data_filtered_top_20_metabolites.R
#data used here was generated by analysis of MSI output in 06.16.21_combinding_TDpointonefive_flux_values_from_MSI_runs.py 

#further based on Date_TBA_microbial_flux_analysis_Burns_data_filtered_top_20_flux_pathways.R, with updates for Hale data
#furtherer based on 09.08.21_Hale2018_fluxpath_analysis.R; modification here is that we're looking at subsystems, rather than just fluxpaths.

##########Preliminary steps
rm(list = ls()) #clear out environment as needed. 
setwd("/Users/adamo010/Documents/MICOM_CRC_FBA/Individual_specific_microbiomes/Analyzing_MICOM_communities_Hale2018/") 

##########Step 1: import data and adjust factors
analyzed_fluxpaths <- read_csv("09.02.21_Hale2018_fluxes_with_metadata_longform.csv")
analyzed_fluxpaths$host_subject_id <- as.factor(analyzed_fluxpaths$host_subject_id)
#colnames(analyzed_metabolites) #print column names as needed

#########NEW Step 2: collapse this dataset by subsystem; average across fluxpaths within a subsystem within a sample.
#drop unnecessary columns
analyzed_fluxpaths_subset <- subset(analyzed_fluxpaths, select = -c(...1, fluxpath_description, fluxpath_formula, sampleIDs_from_read_counts))
#create a new sorting column to get unique values for each Sample_ID/Fluxpath_subsystem combo
analyzed_fluxpaths_subset <- analyzed_fluxpaths_subset %>% 
  mutate(sorting_col = paste0(Sample_ID, "_", fluxpath_subsystem)) 
#AAAA I think I have to replace NANs with 0s. Otherwise the averages won't work 
#analyzed_fluxpaths_subset$fluxpath_amount[is.na(analyzed_fluxpaths_subset$fluxpath_amount)] = 0
#hmmmm, is it better to convert to zeroes or drop??? Probably drop. Let's go with that.
analyzed_fluxpaths_subset_noNANs <- analyzed_fluxpaths_subset[!is.na(analyzed_fluxpaths_subset$fluxpath_amount),]

#then average across sorting col:
analyzed_fluxpaths_subset_averaged <- analyzed_fluxpaths_subset_noNANs %>%        # Specify data frame
  group_by(sorting_col, Sample_ID, normal_adjacent_or_tumor_tissue_specimen, host_subject_id, fluxpath_subsystem) %>%    # Specify group indicator (columns to keep)
  summarise_at(vars(fluxpath_amount),                   # Specify column
               list(av_fluxpath_amount = mean))         # Specify function
#nice. End up with a dataframe where each row is a unique subsystem/sample average of fluxes from pathways within that subsystem

#########Step 3: filter by flux subsystems which are active in at least half the total number of samples 
#Filter to only include flux pathways which are active in more than half of the samples, which translates to... 
#number of times a value appears (at all) in fluxpath_name has to be greater than or equal to 44
#first, create a filtering dataframe that contains counts of each fluxpath_name incidence in the dataframe
analyzed_fluxpaths_filtering_df <- analyzed_fluxpaths_subset_averaged %>%  #create a new dataframe by filtering the old one
  group_by(fluxpath_subsystem) %>% #group by the column of interest
  summarize(n = n_distinct(av_fluxpath_amount)) #count the number of occurrences of each unique value in metabolite_amount and store as variable 'n'

#subtracting 1 b/c nan is counted as a unique value
#create a new dataframe that merges the old and new dataframes
analyzed_fluxpaths_filter_added <- dplyr::inner_join(analyzed_fluxpaths_subset_averaged, analyzed_fluxpaths_filtering_df, by= "fluxpath_subsystem")
#now, remove all rows where n (the name of the count column) is less than 88. Going with 100% coverage!
analyzed_fluxpaths_filtered <- filter(analyzed_fluxpaths_filter_added, n == 92)
rm(analyzed_fluxpaths_filtering_df, analyzed_fluxpaths_filter_added)

#########Step 3: filter to only include paired taxa (i.e. taxa which appear in both tumor and normal samples)
#first, drop all rows that contain nan in metabolite_amount column: already done!
#analyzed_fluxpaths_filtered_noNANs <- analyzed_fluxpaths_filtered[!is.na(analyzed_fluxpaths_filtered$fluxpath_amount),]
#then, do some reordering- maybe this is just a Hale data thing? Get host_subject_id next to fluxpath_subsystem
#analyzed_fluxpaths_filtered_noNANs <- analyzed_fluxpaths_filtered_noNANs[, c(1,2,3,9,4,5,6,7,8,10)] #this is edited from previous version, as my output is a bit different

#then, create a new dataframe where only taxa with pathway activity in BOTH tumor and normal samples (i.e. Patient_Blind_id appears twice) are included
#create a sorting column that unites fluxpath name and Patient_Blind_ID
analyzed_fluxpaths_filtered$sorting_col2 <- paste(analyzed_fluxpaths_filtered$fluxpath_subsystem, "_", analyzed_fluxpaths_filtered$normal_adjacent_or_tumor_tissue_specimen)
analyzed_fluxpaths_paired_only <- subset(analyzed_fluxpaths_filtered,duplicated(sorting_col2) | duplicated(sorting_col2, fromLast=TRUE))
#nothing got dropped, which makes sense- we only kept subsystems that appeared in all samples, so they must be paired by definition.

##########Step 4: split into multiple dataframes by metabolite and do stats on each metabolite
split_analyzed_fluxpaths_by_flux <- split(analyzed_fluxpaths_paired_only, with(analyzed_fluxpaths_paired_only, interaction(fluxpath_subsystem)), drop = TRUE)
by_flux_stats <- lapply(split_analyzed_fluxpaths_by_flux, function(df){wilcox.test(av_fluxpath_amount~normal_adjacent_or_tumor_tissue_specimen, data=df, exact= FALSE, paired= TRUE)})

##########Step 5: statistics
#make lists of p values
by_flux_pvals <- c() #initialize list
for (elem in by_flux_stats){
  new_value = elem$p.value
  by_flux_pvals <- c(by_flux_pvals, new_value)}
rm(new_value)
#make a list of q-values
by_flux_qvals <- p.adjust(by_flux_pvals, method = "fdr")
#make a list of genus_ids
flux_ids <- c()
for(elem in split_analyzed_fluxpaths_by_flux){
  new_value = elem$fluxpath_subsystem[1]
  flux_ids <- c(flux_ids, new_value)}
#merge all lists together. 
fluxpath_statistics <- data.frame(flux_ids, by_flux_pvals, by_flux_qvals)

###########Step 6: clean up and remove extra variables
rm(elem, split_analyzed_fluxpaths_by_flux, by_flux_pvals, by_flux_qvals, flux_ids, new_value)

###########Step 7: pull out significant p-values only
fluxpath_statistics <- fluxpath_statistics[order(fluxpath_statistics$by_flux_pvals),] #re-order by p-value
top_twenty_fluxpaths <- fluxpath_statistics %>% top_n(-20, by_flux_pvals) #note the -20 here to select the smallest p-values (20 taxa)

##########Step 8: save results as a csv file
write.csv(as.data.frame(top_twenty_fluxpaths), file="09.17.21_filtered_microbial_flux_subsystems_top20_sig_results_Hale2018_data.csv")

##########Step 9: prepare data for graphing 
#LAAAAAAAAAAAAME now I have to calculate differences.
#use analyzed_metabolites_paired_only. First, drop boring cols ALREADY DONE 
analyzed_fluxpaths_paired_wide = subset(analyzed_fluxpaths_paired_only, select = -c(n,sorting_col, sorting_col2, Sample_ID) )
#Split up to wide.
analyzed_fluxpaths_paired_wide <- spread(analyzed_fluxpaths_paired_wide, normal_adjacent_or_tumor_tissue_specimen, av_fluxpath_amount)
#add a new column called diff. tumor-normal
analyzed_fluxpaths_paired_wide$diff <- (analyzed_fluxpaths_paired_wide$tumor - analyzed_fluxpaths_paired_wide$normal)

#NOW we can add a new column with useful binning of differences for graphing purposes
analyzed_fluxpaths_paired_wide <- analyzed_fluxpaths_paired_wide %>%
  mutate(Difference_direction = case_when(
    diff > 0 ~ "tumor > normal",
    diff < 0 ~ "tumor < normal",
    diff == 0 ~ "Zero change",
  ))

write.csv(as.data.frame(analyzed_fluxpaths_paired_wide), file="09.17.21_filtered_microbial_flux_subsystems_top20_data_Hale2018_data.csv")

#step 10: NOW GRAPH??!!!
#Heme synthesis
ggpaired(subset(analyzed_fluxpaths_paired_wide, fluxpath_subsystem %in% c("Heme synthesis")), cond1= "normal", cond2= "tumor",
         fill = "condition", line.color = "Difference_direction") +
  labs(title = expression("Heme synthesis subsystem\n average flux, Hale2018 data"), subtitle = expression("paired Wilcoxan, p=0.0167, q= 0.402")) +
  theme(plot.title = element_text(hjust=0.5), plot.subtitle= element_text(hjust=0.5), plot.margin = unit(c(1,0.1,0.1,0.1), "cm")) +
  scale_y_continuous(name = "Average flux,\nmmol/(gDW * h)", limits=c(0,100)) +
  scale_x_discrete(labels = c('Normal','Tumor')) +
  scale_fill_brewer(palette = "Dark2") +
  scale_color_brewer(palette = "Dark2") +
  theme(legend.position = "bottom", legend.title=element_blank(), legend.text=element_text(size=11)) +
  guides(fill="none")

#Tyrosine metabolism
ggpaired(subset(analyzed_fluxpaths_paired_wide, fluxpath_subsystem %in% c("Tyrosine metabolism")), cond1= "normal", cond2= "tumor",
         fill = "condition", line.color = "Difference_direction") +
  labs(title = expression("Tyrosine metabolism subsystem\n average flux, Hale2018 data"), subtitle = expression("paired Wilcoxan, p=0.0172, q= 0.402")) +
  theme(plot.title = element_text(hjust=0.5), plot.subtitle= element_text(hjust=0.5), plot.margin = unit(c(1,0.1,0.1,0.1), "cm")) +
  scale_y_continuous(name = "Average flux,\nmmol/(gDW * h)", limits=c(0,800)) +
  scale_x_discrete(labels = c('Normal','Tumor')) +
  scale_fill_brewer(palette = "Dark2") +
  scale_color_brewer(palette = "Dark2") +
  theme(legend.position = "bottom", legend.title=element_blank(), legend.text=element_text(size=11)) +
  guides(fill="none")

#Urea cycle
ggpaired(subset(analyzed_fluxpaths_paired_wide, fluxpath_subsystem %in% c("Urea cycle")), cond1= "normal", cond2= "tumor",
         fill = "condition", line.color = "Difference_direction") +
  labs(title = expression("Urea cycle subsystem\n average flux, Hale2018 data"), subtitle = expression("paired Wilcoxan, p=0.0257, q= 0.402")) +
  theme(plot.title = element_text(hjust=0.5), plot.subtitle= element_text(hjust=0.5), plot.margin = unit(c(1,0.1,0.1,0.1), "cm")) +
  scale_y_continuous(name = "Average flux,\nmmol/(gDW * h)", limits=c(0,650)) +
  scale_x_discrete(labels = c('Normal','Tumor')) +
  scale_fill_brewer(palette = "Dark2") +
  scale_color_brewer(palette = "Dark2") +
  theme(legend.position = "bottom", legend.title=element_blank(), legend.text=element_text(size=11)) +
  guides(fill="none")

#Arginine and Proline Metabolism
ggpaired(subset(analyzed_fluxpaths_paired_wide, fluxpath_subsystem %in% c("Arginine and Proline Metabolism")), cond1= "normal", cond2= "tumor",
         fill = "condition", line.color = "Difference_direction") +
  labs(title = expression("Arginine and Proline Metabolism subsystem\n average flux, Hale2018 data"), subtitle = expression("paired Wilcoxan, p=0.0273, q= 0.402")) +
  theme(plot.title = element_text(hjust=0.5), plot.subtitle= element_text(hjust=0.5), plot.margin = unit(c(1,0.1,0.1,0.1), "cm")) +
  scale_y_continuous(name = "Average flux,\nmmol/(gDW * h)", limits=c(0,600)) +
  scale_x_discrete(labels = c('Normal','Tumor')) +
  scale_fill_brewer(palette = "Dark2") +
  scale_color_brewer(palette = "Dark2") +
  theme(legend.position = "bottom", legend.title=element_blank(), legend.text=element_text(size=11)) +
  guides(fill="none")

#tRNA Charging
ggpaired(subset(analyzed_fluxpaths_paired_wide, fluxpath_subsystem %in% c("tRNA Charging")), cond1= "normal", cond2= "tumor",
         fill = "condition", line.color = "Difference_direction") +
  labs(title = expression("tRNA Charging subsystem\n average flux, Hale2018 data"), subtitle = expression("paired Wilcoxan, p=0.0340, q= 0.402")) +
  theme(plot.title = element_text(hjust=0.5), plot.subtitle= element_text(hjust=0.5), plot.margin = unit(c(1,0.1,0.1,0.1), "cm")) +
  scale_y_continuous(name = "Average flux,\nmmol/(gDW * h)", limits=c(0,400)) +
  scale_x_discrete(labels = c('Normal','Tumor')) +
  scale_fill_brewer(palette = "Dark2") +
  scale_color_brewer(palette = "Dark2") +
  theme(legend.position = "bottom", legend.title=element_blank(), legend.text=element_text(size=11)) +
  guides(fill="none")


