library("ggplot2")
library("readxl")
library("dplyr")
library("tidyverse")
library("ggpubr")
library("ape")
library("lme4")
library("gplots")
library("plotly")
library("tidyr")
library("vegan")
library("data.table")
library("stringr")

#the goal here is to calculate p-values for the significant differences between flux pathways to use for correlating with gene expression
#differences. Also, we want to abundance filter flux pathways that do not appear in at least half the number of samples 
#(before running stats on identifying fluxes with significantly different activity between tumor and normal samples).

#loosely based on 05.27.21_microbial_inputoutput_analysis_Burns_data_filtered_top_20_metabolites.R
#data used here was generated by analysis of MSI output in 06.16.21_combinding_TDpointonefive_flux_values_from_MSI_runs.py 

#further based on Date_TBA_microbial_flux_analysis_Burns_data_filtered_top_20_flux_pathways.R, with updates for Hale data
#furtherer based on 09.08.21_Hale2018_fluxpath_analysis.R; modification here is that we're looking at subsystems, rather than just fluxpaths.
#and 09.17.21_Hale2018_fluxpath_analysis_at_subsystem_level.R
#and 09.24.21_Hale2018_fluxpath_analysis_at_subsystem_level.R

##########Preliminary steps
rm(list = ls()) #clear out environment as needed. 
setwd("/Users/adamo010/Documents/MICOM_CRC_FBA/Individual_specific_microbiomes/Analyzing_MICOM_communities_Hale2018/") 

##########Step 1: import data and adjust factors
analyzed_fluxpaths <- read_csv("09.02.21_Hale2018_fluxes_with_metadata_longform.csv")
analyzed_fluxpaths$host_subject_id <- as.factor(analyzed_fluxpaths$host_subject_id)
#colnames(analyzed_metabolites) #print column names as needed

#how many unique fluxes are we including here? 
fluxpaths_unique <- unique(analyzed_fluxpaths$fluxpath_name)

#########NEW Step 2: collapse this dataset by subsystem; average across fluxpaths within a subsystem within a sample.
#drop unnecessary columns
analyzed_fluxpaths_subset <- subset(analyzed_fluxpaths, select = -c(...1, fluxpath_description, fluxpath_formula, sampleIDs_from_read_counts))
#create a new sorting column to get unique values for each Sample_ID/Fluxpath_subsystem combo
analyzed_fluxpaths_subset <- analyzed_fluxpaths_subset %>% 
  mutate(sorting_col = paste0(Sample_ID, "_", fluxpath_subsystem)) 
#AAAA I think I have to replace NANs with 0s. Otherwise the averages won't work 
#analyzed_fluxpaths_subset$fluxpath_amount[is.na(analyzed_fluxpaths_subset$fluxpath_amount)] = 0
#hmmmm, is it better to convert to zeroes or drop??? Probably drop. Let's go with that.
analyzed_fluxpaths_subset_noNANs <- analyzed_fluxpaths_subset[!is.na(analyzed_fluxpaths_subset$fluxpath_amount),]

#NEW 09.30.21: take median instead of mean, to control for wild -ve values in some paths. 
#then average across sorting col:
analyzed_fluxpaths_subset_medianed <- analyzed_fluxpaths_subset_noNANs %>%        # Specify data frame
  group_by(sorting_col, Sample_ID, normal_adjacent_or_tumor_tissue_specimen, host_subject_id, fluxpath_subsystem) %>%    # Specify group indicator (columns to keep)
  summarise_at(vars(fluxpath_amount),                   # Specify column
               list(med_fluxpath_amount = median)) %>% # Specify function
  ungroup() #do this so select (dropping columns) can happen later
  
#nice. End up with a dataframe where each row is a unique subsystem/sample average of fluxes from pathways within that subsystem

#how many unique fluxes are we including here? 
fluxpaths_medianed_unique <- unique(analyzed_fluxpaths_subset_medianed$fluxpath_subsystem)

#########Step 3: filter to only include paired taxa (i.e. taxa which appear in both tumor and normal samples)
analyzed_fluxpaths_wide <- select(analyzed_fluxpaths_subset_medianed, -c(sorting_col, Sample_ID)) #drop unnecessary columns
#first, convert to wide form
analyzed_fluxpaths_wide <- analyzed_fluxpaths_wide %>% 
  spread(normal_adjacent_or_tumor_tissue_specimen, med_fluxpath_amount)

#then, drop all rows where both tumor and normal values are nan.
analyzed_fluxpaths_wide_noNANs <- analyzed_fluxpaths_wide[!(is.na(analyzed_fluxpaths_wide$tumor) & is.na(analyzed_fluxpaths_wide$normal)),]
fluxpaths_unique_zerofree <- unique(analyzed_fluxpaths_wide_noNANs$fluxpath_subsystem) #count the number of metabolites
#for this dataset, there are no nans and all data are kept

#replace remaining NA values with zeroes
analyzed_fluxpaths_wide_noNANs$tumor[is.na(analyzed_fluxpaths_wide_noNANs$tumor)] = 0
analyzed_fluxpaths_wide_noNANs$normal[is.na(analyzed_fluxpaths_wide_noNANs$normal)] = 0

#create a difference column
analyzed_fluxpaths_wide_noNANs$difference <- (analyzed_fluxpaths_wide_noNANs$tumor - analyzed_fluxpaths_wide_noNANs$normal)

#remove any rows where differences are 0
analyzed_fluxpaths_wide_noNANs_zerofree <- analyzed_fluxpaths_wide_noNANs[(analyzed_fluxpaths_wide_noNANs$difference !=0),]
#NOTE: for this dataset, a few rows are removed but no fluxpaths

#how many unique fluxes are we including here? 
fluxpaths_zerofree_unique <- unique(analyzed_fluxpaths_wide_noNANs_zerofree$fluxpath_subsystem)

#########Step 3: filter by flux subsystems which are active in at least half the total number of samples 
#first, create a filtering dataframe that contains counts of each fluxpath_name incidence in the dataframe
#need to have n=2 if values are nonzero in both tumor and normal columns; n=1 if values are nonzero in only one column
analyzed_fluxpaths_filtering_df <- analyzed_fluxpaths_wide_noNANs_zerofree %>%  #create a new dataframe by filtering the old one
  mutate(count_of_nonzeros = rowSums(.[3:4]!=0)) %>% #create a new column, count_of_nonzeroes, that counts # nonzero values in columns 3 and 4 for each row
  group_by(fluxpath_subsystem) %>% #group by the column of interest
  summarize(n = sum(count_of_nonzeros)) %>% #sums the count_of_nonzeros values within each fluxpath_subsystem and saves the value as n
  mutate(n) #add n as a new column in the dataframe

#create a new dataframe that merges the old and new dataframes
analyzed_fluxpaths_filter_added <- dplyr::inner_join(analyzed_fluxpaths_wide_noNANs_zerofree, analyzed_fluxpaths_filtering_df, by= "fluxpath_subsystem")
#now, remove all rows where n (the name of the count column) is less than ...
analyzed_fluxpaths_filtered <- filter(analyzed_fluxpaths_filter_added, n >= 92)
#count number of unique metabolites left
fluxpaths_filtered_unique <- unique(analyzed_fluxpaths_filtered$fluxpath_subsystem)

##########Step 4: split into multiple dataframes by metabolite and do stats on each metabolite
#first, edit the dataframe
#drop a couple of columns
analyzed_fluxpaths_paired_only <- select(analyzed_fluxpaths_filtered, -c(difference, n))
#rename a couple of columns to make life easier later. Yeah, yeah, I know, this is the reverse of what I did before.
#analyzed_metabolites_paired_only <- analyzed_metabolites_paired_only %>% rename(normal= mean_genus_GR_normal, tumor= mean_genus_GR_tumor) #rename these columns
#convert to long form
analyzed_fluxpaths_paired_long <- reshape2::melt(data= analyzed_fluxpaths_paired_only,
                                                   id.vars= c("host_subject_id", "fluxpath_subsystem"),
                                                   variable.name = "Description",
                                                   value.name = "fluxpath_activity")
#split up by different metabolites and see where we get. 
split_analyzed_fluxpaths_by_flux <- split(analyzed_fluxpaths_paired_long, with(analyzed_fluxpaths_paired_long, interaction(fluxpath_subsystem)), drop = TRUE)
by_flux_stats <- lapply(split_analyzed_fluxpaths_by_flux, function(df){wilcox.test(fluxpath_activity~Description, data=df, exact= FALSE, paired= TRUE)})

##########Step 5: statistics
#make lists of p values
by_flux_pvals <- c() #initialize list
for (elem in by_flux_stats){
  new_value = elem$p.value
  by_flux_pvals <- c(by_flux_pvals, new_value)}
rm(new_value)
#make a list of q-values
by_flux_qvals <- p.adjust(by_flux_pvals, method = "fdr")
#make a list of genus_ids
flux_ids <- c()
for(elem in split_analyzed_fluxpaths_by_flux){
  new_value = elem$fluxpath_subsystem[1]
  flux_ids <- c(flux_ids, new_value)}
#merge all lists together. 
fluxpath_statistics <- data.frame(flux_ids, by_flux_pvals, by_flux_qvals)

###########Step 6: clean up and remove extra variables
rm(elem, split_analyzed_fluxpaths_by_flux, by_flux_pvals, by_flux_qvals, flux_ids, new_value)

###########Step 7: pull out significant p-values only
fluxpath_statistics <- fluxpath_statistics[order(fluxpath_statistics$by_flux_pvals),] #re-order by p-value

##########Step 8: save results as a csv file
write.csv(as.data.frame(fluxpath_statistics), file="09.30.21_filtered_median_microbial_flux_subsystems_all_results_Hale2018_data.csv")
#new: filter by sig p-values and only keep significantly different p-values
fluxpath_statistics_sig <- fluxpath_statistics[fluxpath_statistics$by_flux_pvals <= 0.05, ]
write.csv(as.data.frame(fluxpath_statistics_sig), file="09.30.21_filtered_median_microbial_flux_subsystems_SIG_results_Hale2018_data.csv")

##########Step 9: prepare data for graphing. NOT DONE HERE
#LAAAAAAAAAAAAME now I have to calculate differences.
#use analyzed_metabolites_paired_only. First, drop boring cols ALREADY DONE 
#analyzed_fluxpaths_paired_wide = subset(analyzed_fluxpaths_paired_long, select = -c(n,sorting_col, sorting_col2, Sample_ID) )
#Split up to wide.
#analyzed_fluxpaths_paired_wide <- spread(analyzed_fluxpaths_paired_wide, normal_adjacent_or_tumor_tissue_specimen, med_fluxpath_amount)
#add a new column called diff. tumor-normal
analyzed_fluxpaths_paired_only$diff <- (analyzed_fluxpaths_paired_only$tumor - analyzed_fluxpaths_paired_only$normal)

#NOW we can add a new column with useful binning of differences for graphing purposes
analyzed_fluxpaths_paired_wide <- analyzed_fluxpaths_paired_only %>%
  mutate(Difference_direction = case_when(
    diff > 0 ~ "tumor > normal",
    diff < 0 ~ "tumor < normal",
    diff == 0 ~ "Zero change",
  ))

write.csv(as.data.frame(analyzed_fluxpaths_paired_wide), file="09.30.21_filtered_median_microbial_flux_subsystems_all_data_Hale2018_data.csv")

#step 10: NOW GRAPH??!!!
freqs <- table(analyzed_fluxpaths_paired_wide$fluxpath_subsystem)
freqs

#tyrosine metabolism
ggpaired(subset(analyzed_fluxpaths_paired_wide, fluxpath_subsystem %in% c("Tyrosine metabolism")), cond1= "normal", cond2= "tumor",
         fill = "condition", line.color = "Difference_direction") +
  labs(title = expression("Tyrosine metabolism subsystem\n median flux, Hale2018 data"), subtitle = expression("paired Wilcoxan, p=0.00590, q= 0.283")) +
  theme(plot.title = element_text(hjust=0.5), plot.subtitle= element_text(hjust=0.5), plot.margin = unit(c(1,0.1,0.1,0.1), "cm")) +
  scale_y_continuous(name = "Median flux,\nmmol/(gDW * h)", limits=c(0,1000)) +
  scale_x_discrete(labels = c('Normal','Tumor')) +
  scale_fill_brewer(palette = "Dark2") +
  scale_color_brewer(palette = "Dark2") +
  theme(legend.position = "bottom", legend.title=element_blank(), legend.text=element_text(size=11)) +
  guides(fill="none")

#Urea cycle
ggpaired(subset(analyzed_fluxpaths_paired_wide, fluxpath_subsystem %in% c("Urea cycle")), cond1= "normal", cond2= "tumor",
         fill = "condition", line.color = "Difference_direction") +
  labs(title = expression("Urea cycle subsystem\n median flux, Hale2018 data"), subtitle = expression("paired Wilcoxan, p= 0.0131, q= 0.315")) +
  theme(plot.title = element_text(hjust=0.5), plot.subtitle= element_text(hjust=0.5), plot.margin = unit(c(1,0.1,0.1,0.1), "cm")) +
  scale_y_continuous(name = "Median flux,\nmmol/(gDW * h)", limits=c(0,850)) +
  scale_x_discrete(labels = c('Normal','Tumor')) +
  scale_fill_brewer(palette = "Dark2") +
  scale_color_brewer(palette = "Dark2") +
  theme(legend.position = "bottom", legend.title=element_blank(), legend.text=element_text(size=11)) +
  guides(fill="none")

#tRNA Charging
ggpaired(subset(analyzed_fluxpaths_paired_wide, fluxpath_subsystem %in% c("tRNA Charging")), cond1= "normal", cond2= "tumor",
         fill = "condition", line.color = "Difference_direction") +
  labs(title = expression("tRNA Charging subsystem\n median flux, Hale2018 data"), subtitle = expression("paired Wilcoxan, p= 0.0340, q= 0.404")) +
  theme(plot.title = element_text(hjust=0.5), plot.subtitle= element_text(hjust=0.5), plot.margin = unit(c(1,0.1,0.1,0.1), "cm")) +
  scale_y_continuous(name = "Median flux,\nmmol/(gDW * h)", limits=c(0,450)) +
  scale_x_discrete(labels = c('Normal','Tumor')) +
  scale_fill_brewer(palette = "Dark2") +
  scale_color_brewer(palette = "Dark2") +
  theme(legend.position = "bottom", legend.title=element_blank(), legend.text=element_text(size=11)) +
  guides(fill="none")

#Alanine and aspartate metabolism
ggpaired(subset(analyzed_fluxpaths_paired_wide, fluxpath_subsystem %in% c("Alanine and aspartate metabolism")), cond1= "normal", cond2= "tumor",
         fill = "condition", line.color = "Difference_direction") +
  labs(title = expression("Alanine and aspartate metabolism subsystem\n median flux, Hale2018 data"), subtitle = expression("paired Wilcoxan, p= 0.0369, q= 0.404")) +
  theme(plot.title = element_text(hjust=0.5), plot.subtitle= element_text(hjust=0.5), plot.margin = unit(c(1,0.1,0.1,0.1), "cm")) +
  scale_y_continuous(name = "Median flux,\nmmol/(gDW * h)", limits=c(0,1000)) +
  scale_x_discrete(labels = c('Normal','Tumor')) +
  scale_fill_brewer(palette = "Dark2") +
  scale_color_brewer(palette = "Dark2") +
  theme(legend.position = "bottom", legend.title=element_blank(), legend.text=element_text(size=11)) +
  guides(fill="none")

#################OLD STUFF
#tRNA Charging
ggpaired(subset(analyzed_fluxpaths_paired_wide, fluxpath_subsystem %in% c("tRNA Charging")), cond1= "normal", cond2= "tumor",
         fill = "condition", line.color = "Difference_direction") +
  labs(title = expression("tRNA Charging subsystem\n average flux, Hale2018 data"), subtitle = expression("paired Wilcoxan, p=0.0340, q= 0.402")) +
  theme(plot.title = element_text(hjust=0.5), plot.subtitle= element_text(hjust=0.5), plot.margin = unit(c(1,0.1,0.1,0.1), "cm")) +
  scale_y_continuous(name = "Average flux,\nmmol/(gDW * h)", limits=c(0,400)) +
  scale_x_discrete(labels = c('Normal','Tumor')) +
  scale_fill_brewer(palette = "Dark2") +
  scale_color_brewer(palette = "Dark2") +
  theme(legend.position = "bottom", legend.title=element_blank(), legend.text=element_text(size=11)) +
  guides(fill="none")


